### PR TITLE
[Mono] Fix offset calculation for nested struct, when pinvoke is enabled

### LINKED ISSUE #90955: [Mono] Regression: amd64 AOT: condition `class1 != ARG_CLASS_NO_CLASS' not met
### Description

Unchanged code from .NET 6/7 to .NET 8 no longer compiles against MonoAOT.

### Reproduction Steps

1. Use the following .NET 5 artefact: [Silk.NET.SDL.dll.changetheextension.txt](https://github.com/dotnet/runtime/files/12413100/Silk.NET.SDL.dll.changetheextension.txt) (or you can attempt to use the latest Silk.NET.SDL NuGet artefact)
2. Attempt to compile using Mono on amd64
3. Observe the assertion fail

### Expected behavior

Compiles as it does with .NET 6/7 Mono.

### Actual behavior

Does not compile.

Full error logs:
```
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Precompiling failed for C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidInputDemo\obj\Release\net8.0-android\android-x64\linked\Silk.NET.SDL.dll with exit code -1073740791. [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidInputDemo\AndroidInputDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Mono Ahead of Time compiler - compiling assembly C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidInputDemo\obj\Release\net8.0-android\android-x64\linked\Silk.NET.SDL.dll [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidInputDemo\AndroidInputDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : AOTID A3075E1A-B18B-3031-4C59-00EC193478B1 [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidInputDemo\AndroidInputDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Using profile data file 'C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\dotnet.aotprofile' [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidInputDemo\AndroidInputDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Added 0 methods from profile. [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidInputDemo\AndroidInputDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : * Assertion at D:\a\_work\1\s\src\mono\mono\mini\mini-amd64.c:760, condition `class1 != ARG_CLASS_NO_CLASS' not met [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidInputDemo\AndroidInputDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Precompiling failed for C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidDemo\obj\Release\net8.0-android\android-x64\linked\Silk.NET.SDL.dll with exit code -1073740791. [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidDemo\AndroidDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Mono Ahead of Time compiler - compiling assembly C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidDemo\obj\Release\net8.0-android\android-x64\linked\Silk.NET.SDL.dll [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidDemo\AndroidDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : AOTID 214C00ED-E566-A415-E0D8-46AB9AC076CB [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidDemo\AndroidDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Using profile data file 'C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\dotnet.aotprofile' [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidDemo\AndroidDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Added 0 methods from profile. [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidDemo\AndroidDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : * Assertion at D:\a\_work\1\s\src\mono\mono\mini\mini-amd64.c:760, condition `class1 != ARG_CLASS_NO_CLASS' not met [C:\runner\_work\Silk.NET\Silk.NET\examples\CSharp\OpenGL Demos\AndroidDemo\AndroidDemo.csproj]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Precompiling failed for C:\runner\_work\Silk.NET\Silk.NET\src\Lab\Experiments\TriangleNET6\obj\Release\net8.0-android\android-x64\linked\Silk.NET.SDL.dll with exit code -1073740791. [C:\runner\_work\Silk.NET\Silk.NET\src\Lab\Experiments\TriangleNET6\TriangleNET6.csproj::TargetFramework=net8.0-android]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Mono Ahead of Time compiler - compiling assembly C:\runner\_work\Silk.NET\Silk.NET\src\Lab\Experiments\TriangleNET6\obj\Release\net8.0-android\android-x64\linked\Silk.NET.SDL.dll [C:\runner\_work\Silk.NET\Silk.NET\src\Lab\Experiments\TriangleNET6\TriangleNET6.csproj::TargetFramework=net8.0-android]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : AOTID 3D729075-93D5-64C6-1763-7E3C129EC501 [C:\runner\_work\Silk.NET\Silk.NET\src\Lab\Experiments\TriangleNET6\TriangleNET6.csproj::TargetFramework=net8.0-android]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Using profile data file 'C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\dotnet.aotprofile' [C:\runner\_work\Silk.NET\Silk.NET\src\Lab\Experiments\TriangleNET6\TriangleNET6.csproj::TargetFramework=net8.0-android]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : Added 0 methods from profile. [C:\runner\_work\Silk.NET\Silk.NET\src\Lab\Experiments\TriangleNET6\TriangleNET6.csproj::TargetFramework=net8.0-android]
  [ERR] Pack: C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Microsoft\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.0-preview.6.359\targets\Microsoft.Android.Sdk.Aot.targets(109,5): error : * Assertion at D:\a\_work\1\s\src\mono\mono\mini\mini-amd64.c:760, condition `class1 != ARG_CLASS_NO_CLASS' not met [C:\runner\_work\Silk.NET\Silk.NET\src\Lab\Experiments\TriangleNET6\TriangleNET6.csproj::TargetFramework=net8.0-android]
  [ERR] Pack: Target Pack has thrown an exception
```

### Regression?

Yes, these applications compiled on .NET 7

### Known Workarounds

No known workarounds.

### Configuration

Running on Windows Server 2022 using workload 34.0.0-preview.6.359. Attempting to compile for amd64. It is specific to that configuration as no other configurations failed. CI job: https://github.com/dotnet/Silk.NET/actions/runs/5933030128/job/16087847725?pr=1591

### Other information

Perhaps it is indeed some undefined behaviour exerted by our bindings, but without knowing which specific function is exercising that undefined behaviour and causing Mono to crash, I don't know what I can do to work around this.

### PR DESCRIPTION
**Pull request by fanyang-mono**
Fixes: #90955

This is an existing bug for both JIT and AOT. Basically, the issue was that the offset from one layer above wasn't carried on when calculating offset for nested struct fields. I've added a test to lock down this behavior and verified that the test failed without this change.

### CODE REVIEW

**Comment by fanyang-mono** [comment]
This should be backported to .NET8/7/6.

**Review by lambdageek — APPROVED** [review]

**Review by lambdageek — COMMENTED** [review]
The test can be simplified - you don't need any methods in the struct definitions - just fields.

The test can also go into `src/tests/Interop/StructMarshalling/PInvoke/` and can be updated like the other .csproj files in that directory to use the `.cpp` file to implement a C++ function that takes the right kind of argument struct.

File: src/tests/JIT/Directed/pinvoke/GameControllerButtonBind.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            if (bindType is not null)
+            {
+                BindType = bindType.Value;
+            }
+
+            if (value is not null)
+            {
+                Value = value.Value;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (fanyang-mono) [reply src/tests/JIT/Directed/pinvoke/GameControllerButtonBind.cs]: Cleaned up.

**Review by lambdageek — APPROVED** [review]

**Comment by fanyang-mono** [comment]
Newly added test failed on wasm with interpreter. Investigating it now.

**Comment by fanyang-mono** [comment]
The wasm apps created for runtime tests weren't including c++ file/library needed by the test. See details in https://github.com/dotnet/runtime/issues/64127. Disabled newly added test on wasm.

**Comment by fanyang-mono** [comment]
Library test failures were related to https://github.com/dotnet/runtime/issues/91410.

**Pull request merged**